File: src/main/java/com/hsmchurch/app/noticeboard/application/NoticeService.java
package com.hsmchurch.app.noticeboard.application;

import com.hsmchurch.app.common.exceptions.NotFoundException;
import com.hsmchurch.app.noticeboard.domain.Notice;
import com.hsmchurch.app.noticeboard.domain.NoticeRepository;
import com.hsmchurch.app.noticeboard.ui.request.NoticeUpdateRequest;
import com.hsmchurch.app.noticeboard.ui.request.NoticeUploadRequest;
import com.hsmchurch.app.noticeboard.ui.response.NoticeResponse;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;

import java.time.LocalDateTime;
import java.util.List;

import static com.hsmchurch.app.common.support.CrudStringFormat.DELETE_FAIL;

@Slf4j
@Service
@Transactional
@RequiredArgsConstructor
public class NoticeService {

    private final static String ENTITY_NAME = "게시글";
    private final NoticeRepository noticeRepository;

    public Notice upload(final NoticeUploadRequest noticeUploadRequest,
                         final Long writerId) {
        return noticeRepository.save(Notice.of(noticeUploadRequest, writerId));
    }

    public Notice findById(final Long noticeId) {
        return noticeRepository.findById(noticeId)
                .orElseThrow(() -> new NotFoundException(ENTITY_NAME, noticeId));
    }

    public NoticeResponse findById2(final Long noticeId) {
        return noticeRepository.findOneWithWriterName(noticeId);
    }

    public Notice updateNotice(final NoticeUpdateRequest noticeUpdateRequest,
                               final Long boardId,
                               final Long writerId) {
        final Notice notice = findById(boardId).update(noticeUpdateRequest, writerId);
        return noticeRepository.save(notice);
    }

    public boolean delete(final Long boardId) {
        try {
            findById(boardId).markAsDeleted();
            return true;
        } catch (NotFoundException e) {
            log.error(DELETE_FAIL.apply(ENTITY_NAME), boardId, e);
            return false;
        }
    }

    public boolean delete(final Long boardId,
                          final Long writerId) {
        try {
            return findById(boardId).checkAndDelete(writerId);
        } catch (NotFoundException e) {
            log.error(DELETE_FAIL.apply(ENTITY_NAME), boardId, e);
            return false;
        }
    }

    public Page<NoticeResponse> findListAll(final Pageable pageable) {
        return noticeRepository.findListAll(pageable);
    }

}

File: src/main/java/com/hsmchurch/app/noticeboard/application/NoticeService.kt
package com.hsmchurch.app.noticeboard.application

import com.hsmchurch.app.common.exceptions.NotFoundException
import com.hsmchurch.app.common.support.CrudStringFormat.DELETE_FAIL
import com.hsmchurch.app.common.support.logger
import com.hsmchurch.app.noticeboard.domain.Notice
import com.hsmchurch.app.noticeboard.domain.NoticeRepository
import com.hsmchurch.app.noticeboard.ui.request.NoticeUpdateRequest
import com.hsmchurch.app.noticeboard.ui.request.NoticeUploadRequest
import com.hsmchurch.app.noticeboard.ui.response.NoticeResponse
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import javax.transaction.Transactional

@Service
@Transactional
open class NoticeService constructor(
        private val noticeRepository: NoticeRepository
) {
    companion object {
        private val ENTITY_NAME = "게시글"
        private val log = logger()
    }

    fun findById(noticeId: Long): Notice = noticeRepository.findByIdOrNull(noticeId)
            ?: throw NotFoundException(ENTITY_NAME, noticeId)

    fun findById2(noticeId: Long): NoticeResponse = noticeRepository.findOneWithWriterName(noticeId)
            ?: throw throw NotFoundException(ENTITY_NAME, noticeId)

    fun findListAll(pageable: Pageable): Page<NoticeResponse?> = noticeRepository.findListAll(pageable)

    fun upload(noticeUploadRequest: NoticeUploadRequest,
               writerId: Long): Notice = noticeRepository.save(Notice.of(noticeUploadRequest, writerId))

    fun updateNotice(noticeUpdateRequest: NoticeUpdateRequest,
                     boardId: Long,
                     writerId: Long): Notice {
        val notice: Notice = findById(boardId).update(noticeUpdateRequest, writerId)
        return noticeRepository.save(notice)
    }

    fun delete(boardId: Long): Boolean {
        return try {
            findById(boardId).markAsDeleted()
            true
        } catch (e: NotFoundException) {
            log.error(DELETE_FAIL(ENTITY_NAME), boardId, e)
            false
        }
    }

    fun delete(boardId: Long,
               writerId: Long): Boolean {
        return try {
            findById(boardId).checkAndDelete(writerId)
        } catch (e: NotFoundException) {
            log.error(DELETE_FAIL(ENTITY_NAME), boardId, e)
            false
        }
    }

}
